Geodata column and Map View (WIP)

flisowna opened the pull request:
## Change Summary

MVP for adding a GeoData column type and Map View, so a user can show data which have geographical aspect. This includes: 

- Ability to Add and Edit GeoData columns 
- Ability create, edit and share a Map View, which shows table data on a leaflet.js map based on the selected geo data column
- For now a user can insert directly the longitude and latitude values or set the geodata on the current location of a device
- The max number of rows shown on a map is 1000
- The whole feature is at the moment only visible after explicitly activated by the user: 
  - for that, click three times on the area on the right side of the 'Views' button; then a few new symbols will appear
    <img width="530" alt="image" src="https://user-images.githubusercontent.com/35769846/215346789-fffb41b2-9eb3-4fe2-a4df-bf5dea4a3c4e.png">
  - then, click on the map marker symbol
  - you are now able to see the Geo Data column type and the button for creating a new Map View

## Out of scope for this MVP PR
- Only Open Street Map as a data provider for now (Google Maps and other are potential candidates for next steps) 
- Address support (only latitute/longitute is supported for now)
- Automatic zoom and moving to the map area which shows all markers
- support for tables with more than 1000 rows

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification
TBD

## Additional information / screenshots (optional)

Anything for maintainers to be made aware of

spaudanjo commented on packages/nc-gui/composables/useMapViewDataStore.ts:
```diff
@@ -0,0 +1,162 @@
+  ) => {
+    const formattedData = ref<Row[]>([])
+
+    const { api } = useApi()
```
maybe also more a question to the NocoDB team: what are the conventions about when to use `useApi` and when `const { $api } = useNuxtApp()`. My understanding is that when calling useApi, it creates a new instance (if it's not called with the parameter `useGlobalInstance=true`) - but in which cases are new instances for the API really necessary? 

In this case I think you can quickly validate/test the assumption that we can just use `$api` (and get rid of the additional `useApi` call) by changing the code and testing that the map view still works as expected when showing, adding, editing and removing geo data.

wingkwong replied:
> My understanding is that when calling useApi, it creates a new instance (if it's not called with the parameter useGlobalInstance=true) - but in which cases are new instances for the API really necessary?

Both would work. With `useApi`, it creates a new clone and you can get the loading status and the error status from the corresponding instance.

spaudanjo commented on packages/nc-gui/composables/useMapViewDataStore.ts:
```diff
@@ -0,0 +1,162 @@
+
+    const geoDataFieldColumn = ref<ColumnType | undefined>()
+
+    const appInfoDefaultLimit = 1000
```
```suggestion
    const defaultPageSize = 1000
```

spaudanjo commented on packages/nc-gui/composables/useMapViewDataStore.ts:
```diff
@@ -0,0 +1,162 @@
+
+    async function insertRow(
+      currentRow: Row,
+      _rowIndex = formattedData.value?.length,
```
Is `_rowIndex` needed?

spaudanjo reviewed (changes requested):
Overall looks pretty good to me :) 

Please see a bunch of inline comments, plus the following comments below. 

Also, maybe you want to 
1) propose here identified test cases for Playwright tests and get the feedback from the NocoDB team on them. Hopefully, after the most crucial tests are added, we can already merge this PR with the Easter Egg feature toggle protection. 
2) put a short screen recording of the features to the PR description

Additional points: 
* There is an empty file packages/nc-gui/components/general/EasterEgg.vue
* GeoData support should be persisted in localStorage
* Open questions to discuss: best to save the geodata in one column, with semicolon separated? Maybe for pragmatic first version yes? 
* Is there a potential way to delete a marker on the map (maybe also a question to the NocoDB team; I was wondering whether an option to delete a record while being in the Edit overlay could be beneficial)? 
* Small UI flaw (can be changed later, when removing the easter egg): the lat and long fields in the cell edit component are not left aligned) 

<img width="223" alt="image" src="https://user-images.githubusercontent.com/249210/210965582-adce3416-48cc-4398-8893-3de697dd2561.png">

flisowna commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,277 @@
+      },
+    })
+  } else {
+    expandedFormClickedLatLongForNewRow.value = clickedLatLongForNewRow
```
yes you're right

flisowna commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,277 @@
+    <div id="mapContainer" ref="mapContainerRef">
+      <a-tooltip placement="bottom" class="tooltip">
+        <template #title>
+          <span v-if="count > 1000"> You're over the limit. </span>
```
I think is still simple to keep it inline.

flisowna replied:
I think is still simple to keep it inline.

flisowna commented on packages/nc-gui/components/smartsheet/sidebar/toolbar/GeodataSwitcher.vue:
```diff
@@ -0,0 +1,14 @@
+<template>
+  <a-tooltip placement="bottomRight">
+    <template #title>
+      <span> Toggle GeoData </span>
```
Yep I think for the easter egg it should be fine.

flisowna commented on packages/nc-gui/context/index.ts:
```diff
@@ -16,6 +16,8 @@ export const ChangePageInj: InjectionKey<ReturnType<typeof useViewData>['changeP
 export const IsGridInj: InjectionKey<Ref<boolean>> = Symbol('is-grid-injection')
 export const IsGalleryInj: InjectionKey<Ref<boolean>> = Symbol('is-gallery-injection')
+export const IsMapInj: InjectionKey<Ref<boolean>> = Symbol('is-map-injection')
+export const IsGeodataActiveInj: InjectionKey<Ref<boolean>> = Symbol('is-geodata-active-injection')
```
nope:)

flisowna commented on packages/nc-gui/composables/useSmartsheetStore.ts:
```diff
@@ -12,8 +12,22 @@ const [useProvideSmartsheetStore, useSmartsheetStore] = useInjectionState(
-    const { sqlUis } = useProject()
+    const { sqlUis, project } = useProject()
+
+    const totalNumberOfTableRows = ref(async () => {
```
correct

flisowna commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,277 @@
+      <a-tooltip placement="bottom" class="tooltip">
+        <template #title>
+          <span v-if="count > 1000"> You're over the limit. </span>
+          <span v-else> You're getting close to the limit. </span>
```
It's corrected now.

flisowna commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,277 @@
+const getMapZoomLocalStorageKey = (viewId: string) => {
+  return `mapView.zoom.${viewId}`
+}
+const getMapCenterLocalStorageKey = (viewId: string) => `mapView.center.${viewId}`
```
I agree

flisowna commented on packages/nocodb/src/lib/models/View.ts:
```diff
@@ -411,7 +432,9 @@ export default class View implements ViewType {
-        } else if (view.type === ViewTypes.KANBAN && !copyFromView) {
+        }
+        // TODO: Check whether/what we need to do here for properly integrate MapView
+        else if (view.type === ViewTypes.KANBAN && !copyFromView) {
```
I added map view, please review it.

wingkwong reviewed (commented):
Good work. Here's some comments after trying out locally. Will take a look at the code part later.

1. What is the difference between `geometry` type and `geoData` type?
2. From your demo video, when you click `Your Location`, it shows `37,5136649`. However, after you save, it shows `37.5136649` in the cell. 
3. May ask Berkay for the advice on the UI modal box (setting lat & lng).
4. Copy `37.5136649` -> click `Set Location` -> paste to either lat or lng -> see the error
5. What would be default position when I first create Map view? Now the default location seems to be somewhere near London. This could be a tricky problem. Let's say if I set all the data in Korea, for better UX, it should show somewhere in Korea covering most points when I open the map. However, calculating the average of lat & lng values may not be accurate since the locations can be all over the world and it could cover 0 points. Overall the idea is to show as many points as possible so that users can see something at least and move less. Probably take a look at Voronoi Diagram or even Convex Hull.

flisowna commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,277 @@
+// import 'vue3-contextmenu/dist/vue3-contextmenu.css'
+import { IsGalleryInj, IsGridInj, IsMapInj, OpenNewRecordFormHookInj, onMounted, provide, ref } from '#imports'
+
+import type { Row as RowType } from '~/lib'
```
Also my linter didn't complain, so I think it's fine

flisowna commented on packages/nc-gui/components/smartsheet/header/CellIcon.vue:
```diff
@@ -0,0 +1,92 @@
+<script setup lang="ts">
```
I removed the .vue file since it was an unused duplicate. Looks like the ts extension was used because no template/stylesheet/script sections were there - it's a pure ts file.

flisowna commented on packages/nc-gui/package.json:
```diff
@@ -50,6 +50,8 @@
     "jsep": "^1.3.6",
     "just-clone": "^6.1.1",
     "jwt-decode": "^3.1.2",
+    "leaflet": "^1.9.2",
```
We added 
* leaflet:  [BSD-2-Clause license](https://github.com/Leaflet/Leaflet/blob/main/LICENSE)
* leaflet.markercluster: [MIT License](https://github.com/Leaflet/Leaflet.markercluster/blob/master/MIT-LICENCE.txt)

flisowna replied:
@o1lab Are all of these licenses fine to be used within NocoDB?

wingkwong replied:
Yes.

flisowna commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,277 @@
+
+provide(IsGalleryInj, ref(false))
+provide(IsGridInj, ref(false))
+provide(IsMapInj, ref(true))
```
I removed it now

flisowna commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,277 @@
+import { latLongToJoinedString } from '~~/utils/geoDataUtils'
+
+provide(IsGalleryInj, ref(false))
+provide(IsGridInj, ref(false))
```
I removed this code now, but your question might still be valid (where exactly is e.g. `IsGridInj` used/read)?

flisowna replied:
maybe @wingkwong knows?

wingkwong replied:
I believe this injection key `IsGridInj` is not in use (globally). This is used when you want to perform some logic on a shared component for a specific view so you need a flag to determine. Currently we have such key but I haven't found any injection. If you are interested, you may search `IsKanbanInj` or `IsFormInj`.

Once you provide a value to this injection key, you can get the value by using `inject` in the child component under `Map.vue`, which I think L13 and L14 are not necessary at all.

flisowna commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,277 @@
+    </div>
+  </div>
+
+  <Suspense>
```
maybe also here @wingkwong can comment? :)

wingkwong replied:
Supposedly yes. However, here we are using `<Suspense>` on `<LazySmartsheetExpandedForm>` which doesn't show initially. From UI perspective, there is not much we can add in a fallback slot.

flisowna commented:
@wingkwong 

Thank you very much for your feedback!

> 1. What is the difference between geometry type and geoData type?

I don't know exactly the specs of geometry column type (since it was there before and I haven't worked on it) but I corrected the wording in the ViewCreate file when I used "geometry" by mistake in a message for a user.

> 2. From your demo video, when you click Your Location, it shows 37,5136649. However, after you save, it shows 37.5136649 in the cell.

It behaves as the decimal column, maybe it depends on the language setting of the user's browser.

> 4. Copy 37.5136649 -> click Set Location -> paste to either lat or lng -> see the error

I fixed it already, thank you for the catch.

> 5. What would be default position when I first create Map view? Now the default location seems to be somewhere near London. This could be a tricky problem. Let's say if I set all the data in Korea, for better UX, it should show somewhere in Korea covering most points when I open the map. However, calculating the average of lat & lng values may not be accurate since the locations can be all over the world and it could cover 0 points. Overall the idea is to show as many points as possible so that users can see something at least and move less. Probably take a look at Voronoi Diagram or even Convex Hull.

I mentioned that as one of the next steps after this PR in Slack. (when creating a new Map View all markers should fit in the view when it gets opened)

I would like to know what is necessary for this PR (reminder: the feature is easter egg protected). Do you think we could implement automatic zoom (point 5) as a follow-up PR?

Please let me know also: 

1. What are critical things still missing in this PR?
2. Which of the tests I mentioned in the PR description are necessary for this PR?

Thank you for your help :)

wingkwong commented:
@flisowna Lemme go through the code once tmr. For point 5, we can leave it first.

flisowna commented:
hey @wingkwong,

No pressure:), but maybe you know (more or less) when you find time to give me some feedback about this PR. This estimation would help me to plan my next steps. 

Of course feel free to write that you can't answer right now this question. 
Hope everything is fine by you.

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,154 @@
+<script lang="ts" setup>
+import type { GeoLocationType } from 'nocodb-sdk'
+import { Modal as AModal, useVModel } from '#imports'
+import { latLongToJoinedString } from '~~/utils/geoDataUtils'
```
We tend to put utils to `#imports`. What you can do is

- Add `export * from './geoDataUtils'` in  `packages/nc-gui/utils/index.ts`
- then import as `import { latLongToJoinedString } from #imports`. In this case, that would be `import { latLongToJoinedString, Modal as AModal, useVModel } from '#imports'`

flisowna replied:
Thanks!

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,154 @@
+  isLoading = true
+  const onSuccess = (position) => {
+    const crd = position.coords
+    console.log('crd', crd)
```
can be removed

flisowna replied:
Ok

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  const onError = (err) => {
+    console.warn(`ERROR(${err.code}): ${err.message}`)
```
if it is error, then `console.error` would be better.

flisowna replied:
done

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,154 @@
+    <a-button>{{ latLongStr }}</a-button>
+    <template #overlay>
+      <a-form :model="formState" class="flex flex-col" @finish="handleFinish">
+        <a-form-item class="inputLat">
```
for the class name, please use something like `foo-bar` instead of `fooBar`.

flisowna replied:
ok

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,154 @@
+            <a-button class="ml-2" @click="onClickSetCurrentLocation">{{ $t('labels.yourLocation') }}</a-button>
+          </div>
+        </a-form-item>
+        <a-form-item class="buttons">
```
prefer better class name

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,154 @@
+          />
+        </a-form-item>
+        <a-form-item>
+          <div style="display: flex; align-items: center; margin-right: 0.5rem">
```
prefer not to have inline style. use class names instead.

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,154 @@
+  @apply ring-transparent;
+}
+
+.inputLat {
+  width: 180px;
+  margin-top: 1rem;
+  margin-bottom: 0.5rem;
+  margin-right: 0.5rem;
+  margin-left: 0.5rem;
+}
+.inputLng {
+  width: 180px;
+  margin-right: 0.5rem;
+  margin-left: 0.5rem;
+}
+
+.fixed-width {
+  width: 24px;
+}
+.buttons {
+  margin-left: auto;
+  margin-bottom: 0;
+}
+.ant-dropdown-menu {
+  height: fit-content;
+  align-items: flex-end;
+}
```
prefer windicss style like `input[type='number']:focus` one

flisowna replied:
Here I'm not sure what you exactly mean can you elaborate?

flisowna replied:
My understanding is that you prefer a very generic, element type focused selector approach, but I think need classes to add element specific adjustments. Can you give a concrete example where you would rewrite a selector?

wingkwong commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,274 @@
+import { ViewTypes } from 'nocodb-sdk'
+import { OpenNewRecordFormHookInj, onMounted, provide, ref } from '#imports'
+import type { Row as RowType } from '~/lib'
+import { latLongToJoinedString } from '~~/utils/geoDataUtils'
```
put into `#imports`

wingkwong commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,274 @@
+    iconCreateFunction(cluster: { getChildCount: () => number }) {
+      return L.divIcon({ html: `${cluster.getChildCount()}`, className: 'geo-map-marker-cluster', iconSize: new L.Point(40, 40) })
+    },
+  })
+  myMap.addLayer(markersClusterGroupRef.value)
```
add line break

wingkwong commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,274 @@
+    if (localStorage != null && mapMetaData?.value?.fk_view_id) {
+      localStorage.setItem(getMapZoomLocalStorageKey(mapMetaData.value.fk_view_id), myMap.getZoom().toString())
+    }
+  })
+  myMap.on('moveend', function () {
```
add line break

wingkwong commented on packages/nc-gui/components/smartsheet/Map.vue:
```diff
@@ -0,0 +1,274 @@
+  </Suspense>
+</template>
+
+<style scoped lang="scss">
+#mapContainer {
+  height: 100vh;
+}
+
+:global(.geo-map-marker-cluster) {
+  background-color: pink;
+  border-radius: 50%;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+}
+</style>
+
+<style>
+.no-underline a {
+  text-decoration: none !important;
+}
+.leaflet-popup-content-wrapper {
+  max-height: 255px;
+  overflow: scroll;
+}
+.tooltip {
+  height: 2rem;
+  max-width: fit-content;
+  position: absolute;
+  top: 10px;
+  right: 10px;
+  padding: 10px;
+  z-index: 500;
+  cursor: default;
+}
+</style>
```
windicss style

flisowna replied:
done

wingkwong commented on packages/nc-gui/components/smartsheet/toolbar/MappedBy.vue:
```diff
@@ -0,0 +1,117 @@
+
+const { loadMapData, loadMapMeta, updateMapMeta, mapMetaData, geoDataFieldColumn } = useMapViewStoreOrThrow()
+
+const stackedByDropdown = ref(false)
```
should it be `mappedByDropdown`?

wingkwong commented on packages/nc-gui/components/smartsheet/toolbar/MappedBy.vue:
```diff
@@ -0,0 +1,117 @@
+
+<template>
+  <a-dropdown v-if="!IsPublic" v-model:visible="stackedByDropdown" :trigger="['click']">
+    <div class="nc-kanban-btn">
```
search `kanban` in this file and replace to `map`

flisowna replied:
done

wingkwong commented on packages/nc-gui/components/smartsheet/toolbar/MappedBy.vue:
```diff
@@ -0,0 +1,117 @@
+  <a-dropdown v-if="!IsPublic" v-model:visible="stackedByDropdown" :trigger="['click']">
+    <div class="nc-kanban-btn">
+      <a-button
+        v-e="['c:kanban:change-grouping-field']"
```
change the name

flisowna replied:
done

wingkwong commented on packages/nc-gui/composables/useMapViewDataStore.ts:
```diff
@@ -0,0 +1,184 @@
+
+    const geoDataFieldColumn = ref<ColumnType | undefined>()
+
+    const defaultPageSize = 1000
+
+    const paginationData = ref<PaginatedType>({ page: 1, pageSize: defaultPageSize })
```
im wondering how to use paginate in a Map view. can you explain it?

flisowna replied:
So the default value of the pagination in nocodb is 25 records. However for the map view it makes sense to show all the records as long it is not causing performance problems. As a safe limit we set 1000 records to be shown on a map view for this MVP and we'll see what is the best solution for the larger data set later. (e.g. one thing Pranav mentioned is to load more records by multiple calls to the BE).

wingkwong reviewed (changes requested):
after the changes, please run `npm run lint` on frontend.

wingkwong commented:
Two more notes.

1. Please also write a summary in a PR. Point form is fine. Just for the record and will be used in release note.
2. When the change requests are resolved, please click `Resolve conversation` to indicate that is resolved.

flisowna commented:
@wingkwong I put the list of potential test candidates that I can think of here now. Please let me know which of these you consider as essential for this MVP PR. 

Please also let me know whether there are other things I still need to do for this MVP PR from your POV. 

### Map View:
 - [ ] Create/delete new map view
 - [ ] Open view
 - [ ] Copy view
 - [ ] Filter works properly
 - [ ] Markers for all rows are shown on the map (IF no filters are applied)
 - [ ] All rows need valid values for the geo data column
 - [ ] Changing primary geo column
 - [ ] Clicking on the marker opens “Edit row” modal
 - [ ] Changes in “Edit row” modal are saved
 - [ ] Saving zoom and center in local storage
 - [ ] Create new marker
 
### Geodata Column
 - [ ] create/edit/delete column
 - [ ] Accepts only valid input for long and lat
 - [ ] “Get current location” sets location correctly
 - [ ] Don’t save numbers if the save button is not clicked

wingkwong commented:
@filiphanes I assume you're talking about Playwright tests. If so, you may take a look at `nocodb/tests/playwright/tests/columnQrCode.spec.ts` for reference. Basic operations (such as creating, editing, deleting map view / geo columns) should be included. Others can be added gradually. However, all items should be tested manually once before merging.

wingkwong commented on packages/nc-gui/composables/useSmartsheetStore.ts:
```diff
@@ -1,7 +1,7 @@
 import type { FilterType, KanbanType, SortType, TableType, ViewType } from 'nocodb-sdk'
 import type { Ref } from 'vue'
-import { computed, ref, unref, useEventBus, useFieldQuery, useInjectionState, useNuxtApp, useProject } from '#imports'
+import { NOCO, computed, ref, unref, useEventBus, useFieldQuery, useInjectionState, useNuxtApp, useProject } from '#imports'
```
seems `NOCO` is not used.

wingkwong commented on packages/nc-gui/composables/useSmartsheetStore.ts:
```diff
@@ -13,7 +13,7 @@ const [useProvideSmartsheetStore, useSmartsheetStore] = useInjectionState(
   ) => {
     const { $api } = useNuxtApp()
-    const { sqlUis } = useProject()
+    const { sqlUis, project } = useProject()
```
seems `project` is not used.

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,155 @@
+  width: 180px;
+}
+
+.input-lat {
+  margin-top: 1rem;
+  margin-bottom: 0.5rem;
+  margin-right: 0.5rem;
+  margin-left: 0.5rem;
+}
+.input-lng {
+  margin-right: 0.5rem;
+  margin-left: 0.5rem;
+}
+
+.btn-group-cancel-submit {
+  margin-left: auto;
+  margin-bottom: 0;
+}
+.ant-dropdown-menu {
+  align-items: flex-end;
+}
+
+.set-location-group {
+  align-items: center;
+  margin-right: 0.5rem;
+}
```
if it only applies to one element, you can just add windicss utility class instead. e.g.

```vue
<a-form-item class="btn-group-cancel-submit">
```

to

```vue
<a-form-item class="ml-auto mb-0">
```

and the following can be removed

```css
.btn-group-cancel-submit {
  margin-left: auto;
  margin-bottom: 0;
}
```

wingkwong reviewed (commented):
1. I was wondering if we really need `Sort` in MAP view. Even thought we sort by some values, the result will be same as Lat/Lng won't be changed. 
2. `Mapped By` is missing in shared view
3. Unticking geo column under `Fields` -> marker still showing. (failed in shared view only)

spaudanjo commented:
Hi @wingkwong 
Regarding `Unticking geo column under Fields -> marker still showing. (failed in shared view only)`
I changed now the behaviour a bit, which also fixes this flaw you pointed out. 
The new behaviour is: When the user changes the `Mapped by` column, this column is set to `show=true` and the UI also disables the checkbox so the user can not un-check it (so they cannot set it to`show=false`). Also, the `Hide all` functionality is now respecting that the `Mapped by` column should not be hidden. 

If you don't see any objections, we try to deliver the Playwright tests you mentioned above until EOW and we hope this PR is then ready to be merged :))

wingkwong reviewed (commented):
@spaudanjo @flisowna It looks good. I got 2 more comments below. You may proceed playwright tests. 

1. I think it'd be nice if the button style is similar to the existing one like attachments (placed in center, font size, you may also consider to use a geo icon inside the button)
![image](https://user-images.githubusercontent.com/35857179/219546990-696e6a23-c107-4a4b-adb7-ab1e15b300a1.png)

2. In share view, expanded form shouldn't be shown after clicking the marker. See other views for reference. I think in the future, we can implement tooltips on the map. Like hovering the marker will show the card including the record data. 
![image](https://user-images.githubusercontent.com/35857179/219548112-551e46d4-a3dd-4069-a1e0-968f80eaf77f.png)

wingkwong commented on tests/playwright/pages/Dashboard/index.ts:
```diff
@@ -42,7 +44,7 @@ export class DashboardPage extends BasePage {
     this.gallery = new GalleryPage(this);
     this.form = new FormPage(this);
-    this.kanban = new KanbanPage(this);
+    this.map = new MapPage(this);
```
here you shouldn't replace kanban :D

flisowna commented:
> @spaudanjo @flisowna It looks good. I got 2 more comments below. You may proceed playwright tests.
> 
> 1. I think it'd be nice if the button style is similar to the existing one like attachments (placed in center, font size, you may also consider to use a geo icon inside the button)
>    ![image](https://user-images.githubusercontent.com/35857179/219546990-696e6a23-c107-4a4b-adb7-ab1e15b300a1.png)
> 2. In share view, expanded form shouldn't be shown after clicking the marker. See other views for reference. I think in the future, we can implement tooltips on the map. Like hovering the marker will show the card including the record data.
>    ![image](https://user-images.githubusercontent.com/35857179/219548112-551e46d4-a3dd-4069-a1e0-968f80eaf77f.png)

Thank you @wingkwong for you feedback. 

I adjusted according to your comment the code and I think it's really important to make the markers (with a tooltip) clickable in share view in the future.

wingkwong commented on tests/playwright/pages/Dashboard/index.ts:
```diff
@@ -10,10 +10,12 @@ import { SettingsPage } from './Settings';
 import { ImportTemplatePage } from './Import/ImportTemplate';
 import { WebhookFormPage } from './WebhookForm';
 import { ProjectsPage } from '../ProjectsPage';
+import { MapPage } from './Map';
```
remove this line - since it has been imported in L13.

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,156 @@
+<script lang="ts" setup>
+import type { GeoLocationType } from 'nocodb-sdk'
+import { Modal as AModal, latLongToJoinedString, useVModel } from '#imports'
+import GeoDataIcon from '~icons/mdi/map-marker'
```
FYI. Icons from mdi are auto imported. Hence, L4 is not necessary since it is used in `<template/>` only.

wingkwong commented on packages/nc-gui/components/cell/GeoData.vue:
```diff
@@ -0,0 +1,156 @@
+      class="group cursor-pointer flex gap-1 items-center mx-auto max-w-32 justify-center active:(ring ring-accent ring-opacity-100) rounded border-1 p-1 shadow-sm hover:(bg-primary bg-opacity-10) dark:(!bg-slate-500)"
+    >
+      <div class="flex items-center gap-2">
+        <GeoDataIcon class="transform dark:(!text-white) group-hover:(!text-accent scale-120) text-gray-500 text-[0.75rem]" />
```
Use `<MdiMapMarker ...`

wingkwong reviewed (commented):
Seems something is wrong in `columnGeoData.spec.ts`. Please take a look.

flisowna commented:
> Seems something is wrong in `columnGeoData.spec.ts`. Please take a look.

@wingkwong I fixed the test. Can you check if there's still something missing or need to be changed?

wingkwong reviewed (changes requested):
Tests are fine. Here's the final change. After that, it's good to merge.

By looking at `develop` branch, the latest migration is up to `nc_025` which means your migration script should start with `nc_026`. In your changes, it includes three migration scripts while the 2nd and 3rd are same. It'd be better to combine them together into one (since we haven't merged to develop branch so it is safe to do so - after merging, your local meta db will be required to recreate).

![image](https://user-images.githubusercontent.com/35857179/220952767-c6f49bbc-05ad-4968-9248-84560da8aa09.png)

Simply put - there will be one migration script called `nc_026_map_view.ts` which include map view & map view column changes. Remember to update `XcMigrationSourcev2.ts` as well.
